### PR TITLE
replace absolute sleep with conditional wait

### PR DESCRIPTION
**Pull request by pstratem**
benchmarks pending...

### CODE REVIEW

**Comment by sipa** [comment]
Agree. If we want the trickling behaviour still, we should make it
conditional on sufficient time having elapsed?

**Comment by pstratem** [comment]
The trickle logic is essentially completely broken as it is.
It really should be fixed with a seperate patch.

**Comment by pstratem** [comment]
with patch on both nodes:
2015-04-05 11:55:36 UpdateTip: new best=000000000000000082ccf8f1557c5d40b21edabb18d2d691cfbf87118bac7254  height=300000  log2_work=78.499549  tx=38463789  date=2014-05-10 06:32:34 progress=0.317754  cache=3218855
2015-04-05 10:55:45 UpdateTip: new best=00000000839a8e6886ab5951d76f411475428afc90947ee320161bbf18eb6048  height=1  log2_work=33.000022  tx=2  date=2009-01-09 02:54:25 progress=0.000000  cache=1

**Comment by ajweiss** [comment]
Is there any substantial change in CPU load for a busy, listening, full
node?

**Comment by pstratem** [comment]
I haven't measured that, but I cant see there being any substantial cpu overhead.

**Comment by pstratem** [comment]
latency_fix both nodes: 3591 seconds
2015-04-05 10:55:45 UpdateTip: new best=00000000839a8e6886ab5951d76f411475428afc90947ee320161bbf18eb6048  height=1  log2_work=33.000022  tx=2  date=2009-01-09 02:54:25 progress=0.000000  cache=1
2015-04-05 11:55:36 UpdateTip: new best=000000000000000082ccf8f1557c5d40b21edabb18d2d691cfbf87118bac7254  height=300000  log2_work=78.499549  tx=38463789  date=2014-05-10 06:32:34 progress=0.317754  cache=3218855

latency_fix sync/master ibd node: 4338 seconds
2015-04-05 22:04:24 UpdateTip: new best=00000000839a8e6886ab5951d76f411475428afc90947ee320161bbf18eb6048  height=1  log2_work=33.000022  tx=2  date=2009-01-09 02:54:25 progress=0.000000  cache=1
2015-04-05 23:16:32 UpdateTip: new best=000000000000000082ccf8f1557c5d40b21edabb18d2d691cfbf87118bac7254  height=300000  log2_work=78.499549  tx=38463789  date=2014-05-10 06:32:34 progress=0.317444  cache=1728437

master sync/master ibd: 4952
2015-04-06 00:16:39 UpdateTip: new best=00000000839a8e6886ab5951d76f411475428afc90947ee320161bbf18eb6048  height=1  log2_work=33.000022  tx=2  date=2009-01-09 02:54:25 progress=0.000000  cache=1
2015-04-06 01:39:11 UpdateTip: new best=000000000000000082ccf8f1557c5d40b21edabb18d2d691cfbf87118bac7254  height=300000  log2_work=78.499549  tx=38463789  date=2014-05-10 06:32:34 progress=0.317379  cache=2357152

**Comment by gmaxwell** [comment]
That is about a 28% speedup. Awesome.  Care to get started on a separate PR based on this which improves the trickle logic?

**Comment by sipa** [comment]
lightly-tested-ACK

**Comment by ajweiss** [comment]
I've been playing with this a little bit and it looks pretty good.  I ran under a test harness that simulated 125 clients that sent 1-2 messages/sec each (ping/pongs) in order to take a look at changes in the load profile and it does roughly double the overhead, but on my system (2014 era decent consumer PC) I only see about 6% CPU utilization jumping to about 12% so it's probably negligible.  (This doubling seems pretty constant even under much higher load, so something to keep in mind for the future.)  I'm also playing with running this as a listening node with so-far, so-good results.  These sorts of changes tend to scare me a bit as they can have nonobvious network effects, so I'm going to keep playing and will report if I find anything...

**Comment by gmaxwell** [comment]
We can likely back off the timeout in the future, the current value is probably excessively conservative.

**Comment by ajweiss** [comment]
Wait, wait.  It looks like this pull is completely subsumed by #5989?  I suppose they need to go in together anyway...

**Comment by gmaxwell** [comment]
@ajweiss it's just building on it.

File: src/net.cpp [hunk 1]
```diff
@@ -1417,7 +1423,7 @@ void ThreadMessageHandler()
         }
 
         if (fSleep)
-            MilliSleep(100);
+            messageHandlerCondition.timed_wait(lock, boost::posix_time::microsec_clock::universal_time()+boost::posix_time::milliseconds(100));
```
Comment (sipa) [inline src/net.cpp]: Spaces around +

**Comment by gavinandresen** [comment]
ACK. Code looks good (really tiny nit: messageHandlerCondition could be declared static).
Runs nicely on OSX, still uses ~no CPU when there is nothing to do.

**Comment by laanwj** [comment]
Code changes look good to me, going to test this.

**Comment by sipa** [comment]
ACK. Squash please?

**Comment by pstratem** [comment]
Squashed

**Comment by laanwj** [comment]
Tested ACK

**Comment by rebroad** [comment]
Does fSleep still need to be conditional on !vRecvMsg.empty() and vRecvMsg[0].complete() given the new code added?

**Pull request merged**